active-rfcs/0000-build-performance.md

```diff
@@ -0,0 +1,187 @@
+2. In the compiler, mark the component as having statically included a directive.
+  - How to mark is up to the implementer, but we have other metadata attached to hydrated component usage already, and it would make sense to follow this same method.
+3. When rendering, if a component contains a client directive, make sure the directive is matched by the marking in __(2)__.
+4. If the marking is not there, we know that the directive must have been added statically. Throw an `Error` message for this, letting the user know that the directive must be added statically.
+
+## The `local:` directive
+
+A new `local:` directive will be introduced that can be used to set relative links on any element. The syntax looks like this:
+
+```astro
+<img local:src="../images/penguin.png" />
+```
+
+The implementation is:
+
+1. The compiler will see the `local:` directives. They must be used *statically* and not added dynamically (they will be ignored if added dynamically).
+1. The compiler will convert the value of the directive into an import statement. For example the above becomes:
+
+    ```astro
+    ---
+    import imgUrl from "../images/penguin.png";
+    ---
+    <img local:src={imgUrl} />
+    ```
+1. If the value of the directive is *not* a static string, the compiler should ignore that usage and not add an import statement.
+1. The compiler include a warning about the non-static usage of the `local:` directive.
+
+There are a couple of special cases that need to be accounted for:
+
+- `<img srcset>` value is more complex than just a single URL. We need to parse the value and add multiple import statements.
+- A `<link rel=stylesheet>` tag will resolve in an import statement that is side-effectual. As such, we should *remove* the actual link tag in the compiler to prevent it from being rendered.
+
+## Deprecate Astro.resolve
+
+To deprecate `Astro.resolve` we should:
+
+- Add a warning to the `Astro.resolve` method that says that it is deprecated and links to documentation on alternatives such as the `local:` directive and `import.meta.glob`.
+- While the feature is deprecated but not removed update the build process to *directly* copy anything that's resolved by `Astro.resolve`, by overloading the usage of the method during the build.
+  - This will allow users who have build sites dependant on the feature to migrate slowly.
+- After one major version of Astro, replace the warning with an error, preventing its usage in dev or the build.
+
+# Drawbacks
+
+- We have heavily promoted the usage of `Astro.resolve` since its inception, as a way to build assets contained in `src/`. This will be a big departure.
+- Having `client:` and `local:` directives both be static might be unintuitive to some, since regular attributes can be added dynamically.
+  - In a future RFC it might make sense to have a blanket requirement that *all* directives be statically added to the template. This would teach developers to expect that requirement.
```

Comment:
big +1 for this, I like it as a rule since directives are explicitly "a thing for the Astro compiler/runtime to handle" so it makes sense that they are as easy to detect as possible.